### DISCUSSION TITLE
Any ideas on why there are large differences when regridding E3SM land sea mask with xcdat xESMF compared to cdms2 ESMF?

### DISCUSSION THREAD

**Discussion by tomvothecoder** [xCDAT/xcdat#521] [2. Questions] [answered]
### Question criteria

- [X] I added a descriptive title here.
- [X] I searched the [xCDAT GitHub Discussions](https://github.com/xCDAT/xcdat/discussions) to find a similar question and didn't find it.
- [X] I searched the [xCDAT documentation](https://xcdat.readthedocs.io/en/stable/) and/or [Xarray documentation](https://docs.xarray.dev/en/stable/).

### Overview
In e3sm_diags I am replacing the [cdms2 horizontal regridder API](https://cdms.readthedocs.io/en/latest/_modules/cdms2/avariable.html#AbstractVariable.regrid) (ESMF, conservative) with the [xcdat horizontal regridder API ](https://xcdat.readthedocs.io/en/latest/_modules/xcdat/regridder/xesmf.html)(xESMF, conservative). The horizontal regridder is used to regrid a land sea mask to a variable's grid before it is applied on the variable.

### The Problem

**The regridded land sea mask produced by xcdat compared to cdms2 appear to be way off. Since the regridded land sea masks are off, the masked variables are also off.** **Are there any ideas on why this might be happening?**

 _Note, I regridded other variables and the results are relatively close between both regridders. I mainly run into this problem when regridding the land sea mask._

### Implementation Logic

- Input variable file: [cl_Amon_E3SM-2-0_historical_r1i1p1f1_gr_200001-201412.nc](https://esgf-data2.llnl.gov/thredds/fileServer/user_pub_work/CMIP6/CMIP/E3SM-Project/E3SM-2-0/historical/r1i1p1f1/Amon/cl/gr/v20220830/cl_Amon_E3SM-2-0_historical_r1i1p1f1_gr_200001-201412.nc)
- Input land sea mask: [acme_ne30_ocean_land_mask.nc](https://github.com/E3SM-Project/e3sm_diags/blob/main/e3sm_diags/driver/acme_ne30_ocean_land_mask.nc)

1. Regrid an E3SM land sea mask to the grid of an input variable (e.g., "cl")
    - Mask grid: **128 lat x 256 lon**
    - Variable "cl" grid: **180 lat x 360 lon**
2. Update the regridded land sea mask with a lower limit
    - All values below the lower limit are masked
    - The lower limit is the land and ocean region value (e.g., 0.65)
3. Apply the land sea mask to the input variable

### Test Cases (from MCVE section below)

```python
# 1. COMPARE GRIDS -- identical (PASSING)
np.testing.assert_equal(output_grid_xc.lat.data, output_grid_cd.getAxis(0))
np.testing.assert_equal(output_grid_xc.lon.data, output_grid_cd.getAxis(1))

# 2. COMPARE INPUT MASKS -- the same (PASSING)
np.testing.assert_equal(ds_mask_xr["LANDFRAC"].data, mask_tvar.data)

# 3. COMPARE REGRIDDED MASKS -- not equal, regridding results in differences (FAILING)
# FIXME: This test case is failing
np.testing.assert_equal(dv_mas_var, mask_tvar_regrid.data)

# AssertionError:
# Arrays are not equal

# Mismatched elements: 18426 / 64800 (28.4%)
# Max absolute difference: 0.15945476
# Max relative difference: 1.
#  x: array([[[1., 1., 1., ..., 1., 1., 1.],
#         [1., 1., 1., ..., 1., 1., 1.],
#         [1., 1., 1., ..., 1., 1., 1.],...
#  y: array([[[1., 1., 1., ..., 1., 1., 1.],
#         [1., 1., 1., ..., 1., 1., 1.],
#         [1., 1., 1., ..., 1., 1., 1.],...

# 4. COMPARE OUTPUT MASK AVERAGES -- identical (PASSING)
mask_tvar_regrid.mean() == dv_mas_var.mean()

# 5. COMPARE REGRIDDED MASKS WITH LOWER LIMIT -- not equal (FAILING)
# FIXME: This test case is failing
lower_limit = 0.65  # This is based on the region value for land/ocean
mask_var_limit = dv_mas_var < lower_limit
mask_tvar_limit = mask_tvar_regrid < lower_limit

np.testing.assert_equal(mask_var_limit.data, mask_tvar_limit.data)

# AssertionError:
# Arrays are not equal

# Mismatched elements: 261 / 64800 (0.403%)
#  x: array([[[False, False, False, ..., False, False, False],
#         [False, False, False, ..., False, False, False],
#         [False, False, False, ..., False, False, False],...
#  y: array([[[False, False, False, ..., False, False, False],
#         [False, False, False, ..., False, False, False],
#         [False, False, False, ..., False, False, False],...
```

### Are there are any possible answers you came across?

- `xesmf` logic or settings that differ from the cdms2 esmf implementation? 
  - xesmf settings include: periodic, extrap_method, extra_dist_exponent, extra_num_src_pts ([source](https://xcdat.readthedocs.io/en/latest/generated/xcdat.regridder.xesmf.XESMFRegridder.html#xcdat.regridder.xesmf.XESMFRegridder))

### Minimal Complete Verifiable Example (MVCE)

```python
# %%
# flake8 noqa: F401
import cdms2
from cdms2.tvariable import TransientVariable
import numpy as np
import xcdat as xc  # noqa
import xarray as xr


# Source: https://esgf-data2.llnl.gov/thredds/fileServer/user_pub_work/CMIP6/CMIP/E3SM-Project/E3SM-2-0/historical/r1i1p1f1/Amon/cl/gr/v20220830/cl_Amon_E3SM-2-0_historical_r1i1p1f1_gr_200001-201412.nc
FILEPATH = "qa/658-lat-lon-set/cl_Amon_E3SM-2-0_historical_r1i1p1f1_gr_200001-201412.nc"
# Source: https://github.com/E3SM-Project/e3sm_diags/blob/main/e3sm_diags/driver/acme_ne30_ocean_land_mask.nc
LAND_OCEAN_MASK_PATH = "e3sm_diags/driver/acme_ne30_ocean_land_mask.nc"

# %%
# ----------------------------------------------------------------------------
# xCDAT
# ----------------------------------------------------------------------------
# 1. Prepare the input variable and mask datasets.
ds_var = xr.open_dataset(FILEPATH)
ds_mask_xr = xr.open_dataset(LAND_OCEAN_MASK_PATH)

# 2. Regrid the mask variable to the input variable dataset grid.
# NOTE: xesmf "conservative_normed" is equivalent to esmf "conservative"
output_grid_xc = ds_var.regridder.grid
ds_mask_xr_regrid = ds_mask_xr.regridder.horizontal(
    "LANDFRAC", output_grid_xc, tool="xesmf", method="conservative_normed"
)
mask_var_regrid = ds_mask_xr_regrid["LANDFRAC"]

# %%
# ------------------------------------------------------------------------------
# cdms2
# ------------------------------------------------------------------------------
# 1. Prepare the input datasets and variables.
ds_tvar = cdms2.open(FILEPATH)
tvar = TransientVariable(ds_tvar["cl"])

ds_mask_cd = cdms2.open(LAND_OCEAN_MASK_PATH)
mask_tvar = TransientVariable(ds_mask_cd["LANDFRAC"])

# 2. Regrid the mask variable to the input variable dataset grid.
output_grid_cd = tvar.getGrid()
mask_tvar_regrid = mask_tvar.regrid(
    output_grid_cd,
    regridTool="esmf",
    regridMethod="conservative",
)
# ------------------------------------------------------------------------------
# TEST CASES
# ------------------------------------------------------------------------------
# 1. COMPARE GRIDS -- identical (PASSING)
np.testing.assert_equal(output_grid_xc.lat.data, output_grid_cd.getAxis(0))
np.testing.assert_equal(output_grid_xc.lon.data, output_grid_cd.getAxis(1))

# 2. COMPARE INPUT MASKS -- the same (PASSING)
np.testing.assert_equal(ds_mask_xr["LANDFRAC"].data, mask_tvar.data)

# 3. COMPARE REGRIDDED MASKS -- not equal, regridding results in differences (FAILING)
# FIXME: This test case is failing
np.testing.assert_equal(mask_var_regrid, mask_tvar_regrid.data)

# AssertionError:
# Arrays are not equal

# Mismatched elements: 18426 / 64800 (28.4%)
# Max absolute difference: 0.15945476
# Max relative difference: 1.
#  x: array([[[1., 1., 1., ..., 1., 1., 1.],
#         [1., 1., 1., ..., 1., 1., 1.],
#         [1., 1., 1., ..., 1., 1., 1.],...
#  y: array([[[1., 1., 1., ..., 1., 1., 1.],
#         [1., 1., 1., ..., 1., 1., 1.],
#         [1., 1., 1., ..., 1., 1., 1.],...

# 4. COMPARE OUTPUT MASK AVERAGES -- identical (PASSING)
mask_tvar_regrid.mean() == mask_var_regrid.mean()

# 5. COMPARE REGRIDDED MASKS WITH LOWER LIMIT -- not equal (FAILING)
# FIXME: This test case is failing
lower_limit = 0.65  # This is based on the region value for land/ocean
mask_var_limit = mask_var_regrid < lower_limit
mask_tvar_limit = mask_tvar_regrid < lower_limit

np.testing.assert_equal(mask_var_limit.data, mask_tvar_limit.data)

# AssertionError:
# Arrays are not equal

# Mismatched elements: 261 / 64800 (0.403%)
#  x: array([[[False, False, False, ..., False, False, False],
#         [False, False, False, ..., False, False, False],
#         [False, False, False, ..., False, False, False],...
#  y: array([[[False, False, False, ..., False, False, False],
#         [False, False, False, ..., False, False, False],
#         [False, False, False, ..., False, False, False],...

#%%
# PLOT THE RESULTS AND DIFFERENCES
mask_xesmf = mask_var_regrid.copy()
mask_xesmf.plot()

# %%
# Using `from_cdms2`  to turn the TransientVariable into an xr.DataArray
# because it is easier to plot.
mask_esmf = xr.DataArray.from_cdms2(mask_tvar_regrid)
mask_esmf.plot()

# Plot the difference
(mask_xesmf - mask_esmf).plot()

```

### Environment

xcdat=0.6.0rc1
cdms2=3.1.5
xesmf=0.7.1

INSTALLED VERSIONS
------------------
commit: None
python: 3.10.12 | packaged by conda-forge | (main, Jun 23 2023, 22:40:32) [GCC 12.3.0]
python-bits: 64
OS: Linux
OS-release: 3.10.0-1160.76.1.el7.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: ('en_US', 'UTF-8')
libhdf5: 1.14.1
libnetcdf: 4.9.2

xarray: 2023.6.0
pandas: 2.0.3
numpy: 1.24.4
scipy: 1.11.1
netCDF4: 1.6.4
pydap: None
h5netcdf: None
h5py: None
Nio: None
zarr: None
cftime: 1.6.2
nc_time_axis: None
PseudoNetCDF: None
iris: None
bottleneck: 1.3.7
dask: 2023.7.0
distributed: 2023.7.0
matplotlib: 3.7.2
cartopy: 0.21.1
seaborn: None
numbagg: None
fsspec: 2023.6.0
cupy: None
pint: 0.19.2
sparse: 0.14.0
flox: None
numpy_groupies: None
setuptools: 68.0.0
pip: 23.1.2
conda: None
pytest: 7.2.0
mypy: 0.991
IPython: 8.14.0
sphinx: 6.2.1

### Anything else we need to know?

API Documentation:
- [cdms2.avariable.AbstractVariable.regrid docs](https://cdms.readthedocs.io/en/latest/_modules/cdms2/avariable.html#AbstractVariable.regrid)
- [xarray.Dataset.regridder.horizontal_xesmf docs](https://xcdat.readthedocs.io/en/latest/generated/xarray.Dataset.regridder.horizontal_xesmf.html)

**Comment by tomvothecoder** [comment]
Any thoughts or ideas @xCDAT/core-developers?

**Reply by pochedls** [reply]
I think `conservative` in CDMS is `conservative_normed` in xCDAT. Does that resolve the difference?

I'm curious where the difference are. If you plot the difference between the two maps, are the differences along the coasts? How big are the differences?

**Reply by chengzhuzhang** [reply]
I think @pochedls is right, let's try `conservative_normed` first. And looking at difference in maps might help visualize the issue..

**Reply by tomvothecoder** [reply]
Thank you @chengzhuzhang and @pochedls! I just compared the results for regridding with xESMF `conservative` and `conservative_normed`. They both produce identical results (code below).

```python
#%%
# flake8: noqa F401
import numpy as np
import xcdat as xc
import xarray as xr

# Source: https://esgf-data2.llnl.gov/thredds/fileServer/user_pub_work/CMIP6/CMIP/E3SM-Project/E3SM-2-0/historical/r1i1p1f1/Amon/cl/gr/v20220830/cl_Amon_E3SM-2-0_historical_r1i1p1f1_gr_200001-201412.nc
FILEPATH = "qa/658-lat-lon-set/cl_Amon_E3SM-2-0_historical_r1i1p1f1_gr_200001-201412.nc"
# Source: https://github.com/E3SM-Project/e3sm_diags/blob/main/e3sm_diags/driver/acme_ne30_ocean_land_mask.nc
LAND_OCEAN_MASK_PATH = "e3sm_diags/driver/acme_ne30_ocean_land_mask.nc"

# %%
# ----------------------------------------------------------------------------
# xCDAT
# ----------------------------------------------------------------------------
# 1. Prepare the input variable and mask datasets.
ds_var = xr.open_dataset(FILEPATH)
ds_mask_xr = xr.open_dataset(LAND_OCEAN_MASK_PATH)

# 2. Regrid the mask variable to the input variable dataset grid.
output_grid_xc = ds_var.regridder.grid

ds_mask_c = ds_mask_xr.regridder.horizontal(
    "LANDFRAC", output_grid_xc, tool="xesmf", method="conservative"
)
ds_mask_cn = ds_mask_xr.regridder.horizontal(
    "LANDFRAC", output_grid_xc, tool="xesmf", method="conservative_normed"
)

# %%
np.testing.assert_equal(ds_mask_c.LANDFRAC.data, ds_mask_cn.LANDFRAC.data)  # True

#%%
xr.testing.assert_equal(ds_mask_c.LANDFRAC, ds_mask_cn.LANDFRAC) # True
```

I'll plot the maps for the regridded land sea mask to compare the xESMF results against ESMF.

**Reply by tomvothecoder** [reply]
Here's the xESMF documentation comparing the regridding methods: https://xesmf.readthedocs.io/en/latest/notebooks/Compare_algorithms.html

> where `conservative_normed` is just the `conservative` method with the normalization set to `ESMF.NormType.FRACAREA` instead of the default `norm_type=ESMF.NormType.DSTAREA`.

**Reply by tomvothecoder** [reply]
> I'm curious where the difference are. If you plot the difference between the two maps, are the differences along the coasts? How big are the differences?

Here are the plots. From what I can tell, the differences seem to be along the coast with an approximate scale of 0.15 to -0.15.

![landfrac-xesmf-conservative-normed](https://github.com/xCDAT/xcdat/assets/25624127/e0c4713d-e81a-438f-85c3-9cb83efae55d)

![landfrac-esmf-conservative](https://github.com/xCDAT/xcdat/assets/25624127/d7cacc3a-9d0f-4e60-b43e-5717d9db9edf)

![diff-xesmf-esmf](https://github.com/xCDAT/xcdat/assets/25624127/e8162067-c7fd-4005-96ad-f6acdd7eb5e6)

The plotting code extending the MVCE:

```python
#%%
# xESMF
from matplotlib import pyplot as plt

mask_xesmf = mask_var_regrid.copy()
plt.suptitle("xESMF (conservative_normed)")
mask_xesmf.plot()

#%%
# ESMF
mask_esmf = xr.DataArray.from_cdms2(mask_tvar_regrid)
plt.suptitle("ESMF (conservative)")
mask_esmf.plot()

#%%
# Difference
diff = mask_xesmf - mask_esmf
plt.suptitle("Difference (xESMF - ESMF)")
diff.plot()
# %%

```

**Reply by jypeter** [reply]
Some loosely related comments:
- is there an example (or some documentation other than the source code) for doing some easy and clean **regridding with xcdat**? I think I used `regrid2` (or something close) in cdms2, because using xesmf was not clearly documented. `regrid2` was probably the original builtin regridder (written by Bob Drach) but only worked on rectilinear grids
- how does the `xcdat/xesmf` regridding compares with `cdo` regridding? We have often used `cdo` to preprocess the grids, because it's easy and fast, and we were not sure of how to do some trusted regridding in python
- for what it's worth, my colleagues usually recommend a _nearest neighbor_ regridding (`regridnn` in cdo, I think) for land-sea masks
- there are lots of (cdms2-based and other) examples in the old (python2) [vacumm gallery](https://umr-lops-vacumm.ifremer.fr/gallery.html#tutorials). Don't know if some could be used for xcdat based examples. [xoa](https://xoa.readthedocs.io/) is the Python3 version of  a subset of vacumm

Also, I see that when using cdms2, you first open a _file variable_ (i.e pointer) and then convert it to an in-memory variable, using
```
ds_tvar = cdms2.open(FILEPATH)
tvar = TransientVariable(ds_tvar["cl"])
```
This can be done more directly with
```
ds_tvar = cdms2.open(FILEPATH)
tvar = ds_tvar('cl')
```

**Reply by pochedls** [reply]
> is there an example (or some documentation other than the source code) for doing some easy and clean **regridding with xcdat**? I think I used `regrid2` (or something close) in cdms2, because using xesmf was not clearly documented. `regrid2` was probably the original builtin regridder (written by Bob Drach) but only worked on rectilinear grids

Checkout the [horizontal regridding](https://xcdat.readthedocs.io/en/stable/examples/horizontal-regridding.html) section of the xCDAT gallery. 

> how does the `xcdat/xesmf` regridding compares with `cdo` regridding? We have often used `cdo` to preprocess the grids, because it's easy and fast, and we were not sure of how to do some trusted regridding in python

We have not compared xcdat's implementation of `regrid2` or `xESMF` to `cdo`. It seems like `cdo` implements `SCRIP` for (at least [some](https://code.mpimet.mpg.de/projects/cdo/embedded/index.html#x1-6800002.12.3)) regridding operations. It was commented [here](https://github.com/xCDAT/xcdat/issues/357) that SCRIP is no longer maintained. If you have users that use CDO and xCDAT and can comment on how they compare, that would be really helpful (I don't know anyone here at LLNL that uses CDO). 

> for what it's worth, my colleagues usually recommend a _nearest neighbor_ regridding (`regridnn` in cdo, I think) for land-sea masks

Thanks – there are a couple nearest neighbor operations for `xESMF` (and I think the documentation is going to be streamlined in v0.6.0 [currently in [testing](https://github.com/xCDAT/xcdat/discussions/517)] at the end of August). 

> there are lots of (cdms2-based and other) examples in the old (python2) [vacumm gallery](https://umr-lops-vacumm.ifremer.fr/gallery.html#tutorials). Don't know if some could be used for xcdat based examples. [xoa](https://xoa.readthedocs.io/) is the Python3 version of  a subset of vacumm

Thanks – we do have a [gallery](https://xcdat.readthedocs.io/en/stable/gallery.html). xCDAT does not have plotting functionality, but if there are specific utilities that you think could use an example, please let us know.

**Reply by jypeter** [reply]
Thanks for the link to the gallery! I had not checked that example yet, but it looks clear enough

Easy/correct/consistent regridding is very important in our community (especially for multi-model work), and it would not hurt to mention regridding in the [FAQ](https://xcdat.readthedocs.io/en/stable/faqs.html), even if it's only a link to the gallery example. We need as many chances as possible for students to find the right tools when they use Google

I think lots of people at IPSL use `cdo`, or cdo-based scripts. And my understanding is that the regridding part of cdo is a `C` rewrite of the original `Fortran` package. So even if SCRIP is no longer maintained, it's probably not a problem for cdo

@jamesorr and @oliviermarti any comments on xcdat (or esmf) vs other regridding tools?

**Reply by pochedls** [reply]
@jypeter – could you suggest an FAQ? Would it be something like 

* "Do you have real examples of xCDAT in use?" or 
* "Does xCDAT perform regridding?"

**Reply by jypeter** [reply]
@pochedls I have listed some ideas in #524

**Reply by pochedls** [reply]
Thanks @jypeter for documenting/suggesting these documentation improvements!

**Reply by tomvothecoder** [reply]
> Thank you @chengzhuzhang and @pochedls! I just compared the results for regridding with xESMF `conservative` and `conservative_normed`. They both produce identical results (code below).
> 
> ```python
> #%%
> # flake8: noqa F401
> import numpy as np
> import xcdat as xc
> import xarray as xr
> 
> # Source: https://esgf-data2.llnl.gov/thredds/fileServer/user_pub_work/CMIP6/CMIP/E3SM-Project/E3SM-2-0/historical/r1i1p1f1/Amon/cl/gr/v20220830/cl_Amon_E3SM-2-0_historical_r1i1p1f1_gr_200001-201412.nc
> FILEPATH = "qa/658-lat-lon-set/cl_Amon_E3SM-2-0_historical_r1i1p1f1_gr_200001-201412.nc"
> # Source: https://github.com/E3SM-Project/e3sm_diags/blob/main/e3sm_diags/driver/acme_ne30_ocean_land_mask.nc
> LAND_OCEAN_MASK_PATH = "e3sm_diags/driver/acme_ne30_ocean_land_mask.nc"
> 
> # %%
> # ----------------------------------------------------------------------------
> # xCDAT
> # ----------------------------------------------------------------------------
> # 1. Prepare the input variable and mask datasets.
> ds_var = xr.open_dataset(FILEPATH)
> ds_mask_xr = xr.open_dataset(LAND_OCEAN_MASK_PATH)
> 
> # 2. Regrid the mask variable to the input variable dataset grid.
> output_grid_xc = ds_var.regridder.grid
> 
> ds_mask_c = ds_mask_xr.regridder.horizontal(
>     "LANDFRAC", output_grid_xc, tool="xesmf", method="conservative"
> )
> ds_mask_cn = ds_mask_xr.regridder.horizontal(
>     "LANDFRAC", output_grid_xc, tool="xesmf", method="conservative_normed"
> )
> 
> # %%
> np.testing.assert_equal(ds_mask_c.LANDFRAC.data, ds_mask_cn.LANDFRAC.data)  # True
> 
> #%%
> xr.testing.assert_equal(ds_mask_c.LANDFRAC, ds_mask_cn.LANDFRAC) # True
> ```
> 
> I'll plot the maps for the regridded land sea mask to compare the xESMF results against ESMF.

To clarify, I am comparing xESMF conservative vs. xESMF conservative_normed and found they were identical.

This means changing the method does not resolve the large differences with the xESMF results compared to ESMF conservative.

**Reply by pochedls** [reply]
@tomvothecoder – thanks. One other option that would apply in this case: could you try `period=True` in xcdat/xESMF to see if that brings things closer to CDAT?

**Reply by tomvothecoder** [reply]
I tried `periodic=true` but the differences are still the same.

I reran the cdms2 regridder and the first execution produces a warning that `libcf` is being used instead of `esmf`.
**This might be the source of the large differences.**

```python
[/home/vo13/mambaforge/envs/diags_geocat/lib/python3.10/site-packages/cdms2/avariable.py:1289](https://vscode-remote+ssh-002dremote-002bacme1.vscode-resource.vscode-cdn.net/home/vo13/mambaforge/envs/diags_geocat/lib/python3.10/site-packages/cdms2/avariable.py:1289): Warning: 
avariable.regrid: regridTool = 'esmf' but your version does not
seems to be built with esmf, will switch to regridTool = 'libcf'
                  
  warnings.warn(message, Warning)
```

Now I have to figure out how to get cdms2's regridder built with `esmf`. Does anybody have any ideas while I search the docs?

**Reply by tomvothecoder** [reply]
I found an `e3sm_diags` [GitHub issue](https://github.com/E3SM-Project/e3sm_diags/issues/561#issuecomment-1036650724) related to this warning. I think we need to install `esmf` in the conda environment?

I just checked my e3sm_diags development environment and it already includes `esmf=8.4.2` and `esmpy=8.4.2`. Not sure what to try next to get past this warning.

**Reply by chengzhuzhang** [reply]
I was wondering with the same thing. Maybe the issue from e3sm_diags is related. Xylar patched cdms2, so that `esmf` can be properly called by `cdms2`. When I tested, i installed `-c conda-forge/label/cdms2_dev cdms2=3.1.5rc4` that fixed the issue.

**Reply by chengzhuzhang** [reply]
The latest e3sm_diags issue and the fix for `cdms2`.  see here: https://github.com/E3SM-Project/e3sm_diags/issues/711

**Reply by chengzhuzhang** [reply]
Environment related, I'm trying to reproduce, but realized my conda environment for validation (with both xcdat and cdms2 included) is out-dated. It was created early using https://github.com/xCDAT/xcdat-validation/blob/main/conda-env/test_dev.yml.

**Reply by tomvothecoder** [reply, marked as answer]
> I was wondering with the same thing. Maybe the issue from e3sm_diags is related. Xylar patched cdms2, so that `esmf` can be properly called by `cdms2`. When I tested, i installed `-c conda-forge/label/cdms2_dev cdms2=3.1.5rc4` that fixed the issue.

> The latest e3sm_diags issue and the fix for `cdms2`. see here: [E3SM-Project/e3sm_diags#711](https://github.com/E3SM-Project/e3sm_diags/issues/711)

Thank you @chengzhuzhang! I followed your guidance and now the xESMF and ESMF regridding results are close with around 1e-7 absolute and relative max difference. I will mark my comment here as the answer.

To sum up this discussion:

* **Root cause of the large diffs** - `cdms2` was not properly importing `esmf` and using `libcf` as the fallback regridding tool. You will receive this warning below:
  * `Warning: avariable.regrid: regridTool = 'esmf' but your version does not seems to be built with esmf, will switch to regridTool = 'libcf'`
* **Solution** - Use the latest patched `cdms2=3.1.5`, which properly imports `esmf`.
  * `<conda|mamba> update -c conda-forge cdms`
  * Refer to this [PR](https://github.com/conda-forge/cdms2-feedstock/pull/89) for more info on this issue and patch

Code with test cases:

```python
# %%
# flake8 noqa: F401
import cdms2
from cdms2.tvariable import TransientVariable
import numpy as np
import xcdat as xc  # noqa
import xarray as xr


# Source: https://esgf-data2.llnl.gov/thredds/fileServer/user_pub_work/CMIP6/CMIP/E3SM-Project/E3SM-2-0/historical/r1i1p1f1/Amon/cl/gr/v20220830/cl_Amon_E3SM-2-0_historical_r1i1p1f1_gr_200001-201412.nc
FILEPATH = "qa/658-lat-lon-set/cl_Amon_E3SM-2-0_historical_r1i1p1f1_gr_200001-201412.nc"
# Source: https://github.com/E3SM-Project/e3sm_diags/blob/main/e3sm_diags/driver/acme_ne30_ocean_land_mask.nc
LAND_OCEAN_MASK_PATH = "e3sm_diags/driver/acme_ne30_ocean_land_mask.nc"

# %%
# ----------------------------------------------------------------------------
# xCDAT
# ----------------------------------------------------------------------------
# 1. Prepare the input variable and mask datasets.
ds_var = xr.open_dataset(FILEPATH)
ds_mask_xr = xr.open_dataset(LAND_OCEAN_MASK_PATH)

# 2. Regrid the mask variable to the input variable dataset grid.
output_grid_xc = ds_var.regridder.grid
ds_mask_xr_regrid = ds_mask_xr.regridder.horizontal(
    "LANDFRAC", output_grid_xc, tool="xesmf", method="conservative_normed", periodic=True,
)
mask_var_regrid = ds_mask_xr_regrid["LANDFRAC"]

# %%
# ------------------------------------------------------------------------------
# cdms2
# ------------------------------------------------------------------------------
# 1. Prepare the input datasets and variables.
ds_tvar = cdms2.open(FILEPATH)
tvar = TransientVariable(ds_tvar["cl"])

ds_mask_cd = cdms2.open(LAND_OCEAN_MASK_PATH)
mask_tvar = TransientVariable(ds_mask_cd["LANDFRAC"])

# 2. Regrid the mask variable to the input variable dataset grid.
output_grid_cd = tvar.getGrid()
mask_tvar_regrid = mask_tvar.regrid(
    output_grid_cd,
    regridTool="esmf",
    regridMethod="conservative",
)

#%%
# ------------------------------------------------------------------------------
# TEST CASES
# ------------------------------------------------------------------------------
# 3. COMPARE REGRIDDED MASKS -- PASSING
np.testing.assert_equal(mask_var_regrid, mask_tvar_regrid.data)

# AssertionError:
# Arrays are not equal

# Mismatched elements: 3182 / 64800 (4.91%)
# Max absolute difference: 1.1920929e-07
# Max relative difference: 1.4111636e-07
#  x: array([[[1., 1., 1., ..., 1., 1., 1.],
#         [1., 1., 1., ..., 1., 1., 1.],
#         [1., 1., 1., ..., 1., 1., 1.],...
#  y: array([[[1., 1., 1., ..., 1., 1., 1.],
#         [1., 1., 1., ..., 1., 1., 1.],
#         [1., 1., 1., ..., 1., 1., 1.],...

# 5. COMPARE REGRIDDED MASKS WITH LOWER LIMIT -- PASSING
lower_limit = 0.65  # This is based on the region value for land/ocean
mask_var_limit = mask_var_regrid < lower_limit
mask_tvar_limit = mask_tvar_regrid < lower_limit

np.testing.assert_equal(mask_var_limit.data, mask_tvar_limit.data)  # True
```

Difference plot:
![diff-plot](https://github.com/xCDAT/xcdat/assets/25624127/094df0a8-aa82-4f4c-b38f-9d68f8173d8c)

**Comment by chengzhuzhang** [comment]
> > I'm curious where the difference are. If you plot the difference between the two maps, are the differences along the coasts? How big are the differences?
> 
> Here are the plots. From what I can tell, the differences seem to be along the coast with an approximate scale of 0.15 to -0.15.
> 
Checking on some old results comparing cdms (conservative) vs xesmf (conservative_normed). The difference is very small (megabit diff: In [13] in [validation notebook](https://github.com/xCDAT/xcdat-validation/blob/main/validation/v0.3.0/horizontal_regridding/regrid_cdat_xcdat.ipynb).)
